File: c/tree/populating-next-right-pointers-in-each-node-ii.c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */

struct TreeLinkNode *get_next_child(struct TreeLinkNode *root)
{
    while(root){
        if(root->left)
            return root->left;
        if(root->right)
            return root->right;
        root = root->next;
    }

    return NULL;
}

struct TreeLinkNode *get_next(struct TreeLinkNode *root)
{
    while(root && root->left == NULL && root->right == NULL)
        root = root->next;

    return root;
}

void connect(struct TreeLinkNode *root) {
    if(root == NULL)
        return;

    struct TreeLinkNode *parent = get_next(root);
    struct TreeLinkNode *next = get_next_child(root);
    while(parent && next){
        while(parent){
            if(parent->left)
                parent->left->next = parent->right ? parent->right : get_next_child(parent->next);

            if(parent->right)
                parent->right->next = get_next_child(parent->next);

            parent = get_next(parent->next);
        }
        parent = get_next(next);
        next = get_next_child(next);
    }
}

File: python/tree/populating-next-right-pointers-in-each-node-ii.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def get_next(self, root):
        while root and root.left == None and root.right == None:
            root = root.next

        return root

    def get_next_child(self,root):
        parent = self.get_next(root)
        if parent:
            return parent.left if parent.left else parent.right

        return None

    def connect(self, root):
        parent = self.get_next(root)
        next = self.get_next_child(root)

        while parent and next:
            while parent:
                if parent.left:
                    parent.left.next = parent.right if parent.right \
                        else self.get_next_child(parent.next)
                if parent.right:
                    parent.right.next = self.get_next_child(parent.next)

                parent = self.get_next(parent.next)

            parent = self.get_next(next)
            next = self.get_next_child(next)
